packages/sdk/src/webHost.ts

```diff
@@ -3,82 +3,148 @@
 		log.info('app', `Node: ${process.version}`);
 		log.info('app', `${pjson.name}: v${pjson.version}`);
 
-		this._baseDir = options.baseDir || process.env.BASE_DIR;
-		this._baseUrl = options.baseUrl || process.env.BASE_URL;
+		this.options = {
+			baseDir: process.env.BASE_DIR,
+			baseUrl: process.env.BASE_URL,
+			port: process.env.PORT || process.env.port || 3901,
+			mreRoute: '/',
+			...this.options
+		};
+
+		// Azure defines WEBSITE_HOSTNAME
+		if (!this.options.baseUrl && process.env.WEBSITE_HOSTNAME) {
+			this.options.baseUrl = `https://${process.env.WEBSITE_HOSTNAME}`;
+		}
 
-		// Azure defines WEBSITE_HOSTNAME.
-		if (!this._baseUrl && process.env.WEBSITE_HOSTNAME) {
-			this._baseUrl = `https://${process.env.WEBSITE_HOSTNAME}`;
+		// Set the default adapter only if `adapter` was entirely omitted from `options`.
+		// If the caller passed `null` interpret this as "I don't want an adapter added".
+		if (this.options.adapter === undefined) {
+			this.options.adapter = new MultipeerAdapter();
 		}
 
-		// Resolve the port number. Heroku defines a PORT environment var (remapped from 80).
-		const port = options.port || process.env.PORT || 3901;
-
-		// Create a Multi-peer adapter
-		this._adapter = new MultipeerAdapter({ port });
-
-		// Start listening for new app connections from a multi-peer client.
-		this._adapter.listen()
-			.then(server => {
-				this._baseUrl = this._baseUrl || server.url.replace(/\[::\]/, '127.0.0.1');
-				log.info('app', `${server.name} listening on ${JSON.stringify(server.address())}`);
-				log.info('app', `baseUrl: ${this.baseUrl}`);
-				log.info('app', `baseDir: ${this.baseDir}`);
-				if (!!this.baseDir) {
-					this.serveStaticFiles(server);
-				}
-			})
-			.catch(reason => log.error('app', `Failed to start HTTP server: ${reason}`));
-	}
+		// If an adapter was provided, register it.
+		if (this.options.adapter) {
+			this.addAdapter(this.options.mreRoute, this.adapter);
+		}
 
-	private serveStaticFiles(server: Restify.Server) {
-		server.get('/*',
-			// host static binaries
-			(req, res, next) => this.serveStaticBuffers(req, res, next),
-			// host static files
-			Restify.plugins.serveStatic({
-				directory: this._baseDir,
-				default: 'index.html'
+		// Start the web server, enable WebSockets, start serving static files, and start handling WebSocket connections.
+		this._server = Restify.createServer({ name: this.constructor.name });
+		this.wss = new WS.Server({ server: this.server });
+		this.server.listen(this.options.port, () => {
+			this.options.baseUrl = this.options.baseUrl || this.server.url.replace(/\[::\]/, '127.0.0.1');
+			log.info('app', `${this.server.name} listening on ${JSON.stringify(this.server.address())}`);
+			log.info('app', `baseUrl: ${this.baseUrl}`);
+			log.info('app', `baseDir: ${this.baseDir}`);
+			if (!!this.baseDir) {
+				serveStaticFiles();
 			}
-			));
+			handleWebSocketUpgrades();
+		});
+
+		const serveStaticFiles = () => {
```

Comment:
Static file hosting should be served from the adapter host as well. Two apps in the same server aren't going to necessarily merge their public directories.

Reply:
Interesting. For the sake of backward compatibility I'd like to keep it this way for now. Alternately, I can remove support for multiple MREs. It seemed like a neat thing to enable, but this makes it complicated. I don't think WebHost is the right solution for large-scale apps anyway. We shouldn't presume to own the http server or WebSocket server.

---
In reply to: [332689431](https://github.com/microsoft/mixed-reality-extension-sdk/pull/453#discussion_r332689431) [](ancestors = 332689431)

Reply:
I'm in favor of removing multi-MRE support if we can't do it well, or else document the interrelationships.

---
In reply to: [332711372](https://github.com/microsoft/mixed-reality-extension-sdk/pull/453#discussion_r332711372) [](ancestors = 332711372,332689431)

Reply:
Multi-MRE support disabled for now.